Add no-bypass program and improve build system

micheledallerive opened the pull request:
Changes:
- Add UDP-based baseline program with no kernel bypass which can be used as a comparison to the kernel bypass techniques.
- Create a central CMakeLists file which compiles all the subprojects. 
- Move the check of whether the machine is the server host at compile-time, through the flag `__SERVER__`. This can be a small improvement in the critical portions of code where packets are handled, avoiding useless branching and allowing the compiler to do better optimizations.
- Improve command-line arguments (#2) parsing by removing positional arguments and using option arguments. For example, instead of `./pp_poll ens1f1np1 start 1000000 10000 10.10.1.1` the CLI is `./pp_poll -d ens1f1np1 -p 1000000 -i 10000 -s 10.10.1.1`, or `./pp_poll --device ens1f1np1 --packets 1000000 --interval 10000 --server 10.10.1.1`. All programs accept an optional parameter `-m <0,1,2>` or `--measurement <0,1,2>` which specifies the kind of measurement being taken, respectively: all timestamps, only min/max latency or buckets.

micheledallerive committed:
Add "no-bypass" pingpong as a baseline for evaluation.

The no-bypass pingpong uses basic UDP socket programming to run the experiment.
Such a baseline can be useful to evaluate how well bypass techniques work compared to a zero-effort no-bypass technique.

micheledallerive committed:
Modify no-bypass CLI to use getopts.

micheledallerive committed:
Add __SERVER__ compilation flag and arguments parsing methods.

micheledallerive committed:
Add persistence measurement as command line argument.

dvr0 commented on CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+cmake_minimum_required(VERSION 3.10)
+project(det-bypass)
+
+if (NOT DEFINED __SERVER__)
```
What's the reason of using `__SERVER__` instead of  `SERVER`? Aren't double underscore "__var"s reserved for C implementation or similar?

micheledallerive replied:
Honestly there was no specific reason. I changed it to SERVER, which makes it easier to type as well.

dvr0 commented on rdma/README.md:
```diff
@@ -54,19 +54,20 @@ our use case:
 ## Build
 
-The build process follows the same steps as for any CMake project:
+The build process follows the same steps as for any CMake project, with a variable IS_SERVER to indicate if the program
```
forgot to put `__SERVER__` instead of IS_SERVER I guess

dvr0 commented on rdma/CMakeLists.txt:
```diff
@@ -2,11 +2,6 @@ cmake_minimum_required(VERSION 3.10)
 
```
if someone uses the "inner" CMakeFiles it won't complain about `__SERVER__` missing. Hence it would default as a client, right?

micheledallerive replied:
Yea I was assuming the compilation would happen only from the top-level CMake. I isolated the code that handles the configuration variables (i.e. DEBUG and SERVER) and made every sub-cmake import it.

dvr0 reviewed (commented):
Very nice, defining server and client and compile time makes sense and the branching is very well organised. My only concern is, you end up with a program named `ud_pingpong` for instance and you can't tell whether it's a client or a server, the only way you can tell is from the extra options from the usage string.  
Could you maybe add a "hey I am server/client" in the usage? Or something else to differentiate

btw I am being really picky here, the argument "man just recompile if unsure" is legit

micheledallerive committed:
Add better usage message, explicitly stating whether the program is client or server.

micheledallerive committed:
Centralize CMake programs configuration.

micheledallerive committed:
Rename __SERVER__ to SERVER.

micheledallerive committed:
Fix CMake variables setting.

The values that options take are CMake-booleans, i.e. "OFF" and "ON" (i guess?!). Compilation fails because those values are not recognized as valid boolean values in C (of course?!).
